File: 프로그래머스/2/17680. ［1차］ 캐시/［1차］ 캐시.cpp
#include <string>
#include <vector>
#include <deque>
#include <algorithm>

using namespace std;

int solution(int cacheSize, vector<string> cities) {
    int answer = 0;
    
    // 전체 소문자 변환
    for(auto& city : cities){
        transform(city.begin(), city.end(), city.begin(), ::tolower);
    }
    
    // 캐시가 0이면
    if(cacheSize == 0)
        return 5 * cities.size();
    
    int cache_hit = 1;
    int cache_miss = 5;
    int total_time = 0;
    
    deque<string> dq;
    
    for(auto& city : cities){
        auto it = find(dq.begin(), dq.end(), city);
        
        if(it != dq.end()) {
            // 캐시 히트
            total_time += cache_hit;
            dq.erase(it);
            dq.push_back(city);
        } else {
            // 캐시 미스
            total_time += cache_miss;
            if(dq.size() == cacheSize) {
                dq.pop_front();
            }
            dq.push_back(city);
        }
    }
    
    answer = total_time;
    return answer;
}

File: 프로그래머스/2/17680. ［1차］ 캐시/［1차］ 캐시.java
import java.util.*;

class Solution {
    public int solution(int cacheSize, String[] cities) {
        int answer = 0;
        
        if (cacheSize == 0) {
            return cities.length * 5;
        }
        
        Deque<String> dq = new ArrayDeque<>();
        
        for(String city : cities) {
            city = city.toLowerCase();
            
            // 캐시에 있는 경우
            if(dq.remove(city)){
                answer += 1; 
                dq.addLast(city); 
            } else {                
                // 캐시가 가득 찬 경우
                if(dq.size() >= cacheSize){
                    dq.removeFirst();
                }
                answer += 5;
                dq.addLast(city);
            }
        }
        
        return answer;
    }
}